tensordict/tensordict.py

```diff
@@ -4902,6 +4926,78 @@ def _cat(
         return out
 
 
+@implements_for_lazy_td(torch.cat)
+def _lazy_cat(
+    list_of_tensordicts: Sequence[LazyStackedTensorDict],
+    dim: int = 0,
+    out: LazyStackedTensorDict | None = None,
+) -> LazyStackedTensorDict:
+    if not list_of_tensordicts:
+        raise RuntimeError("list_of_tensordicts cannot be empty")
+
+    batch_size = list(list_of_tensordicts[0].batch_size)
+    if dim < 0:
+        dim = len(batch_size) + dim
+    if dim >= len(batch_size):
+        raise RuntimeError(
+            f"dim must be in the range 0 <= dim < len(batch_size), got dim"
+            f"={dim} and batch_size={batch_size}"
+        )
+    stack_dim = list_of_tensordicts[0].stack_dim
+    if any([(td.stack_dim != stack_dim) for td in list_of_tensordicts]):
+        raise RuntimeError("cat lazy stacked tds must have same stack dim")
+
+    batch_size[dim] = sum([td.batch_size[dim] for td in list_of_tensordicts])
+    batch_size = torch.Size(batch_size)
+
+    new_dim = dim
+    if dim > stack_dim:
+        new_dim = dim - 1
+
+    # check that all tensordict match
+    _check_keys(list_of_tensordicts, strict=True)
+    if out is None:
+        out = []
+        if dim == stack_dim:  # if dim is stack, just add all to same list
+            for lazy_td in list_of_tensordicts:
+                out += lazy_td.tensordicts
+        else:
+            for i in range(len(list_of_tensordicts[0].tensordicts)):
+                out.append(
+                    torch.cat(
+                        [lazy_td.tensordicts[i] for lazy_td in list_of_tensordicts],
+                        new_dim,
+                    )
+                )
+        return LazyStackedTensorDict(*out, stack_dim=stack_dim)
+    else:
+        if out.batch_size != batch_size:
+            raise RuntimeError(
+                "out.batch_size and cat batch size must match, "
+                f"got out.batch_size={out.batch_size} and batch_size"
+                f"={batch_size}"
+            )
+        if out.stack_dim != stack_dim:
+            raise RuntimeError(
+                "out.stack_dim and elements stack_sim did not match, "
+                f"got out.stack_dim={out.stack_dim} and stack_dim={stack_dim}"
+            )
+
+        if dim == stack_dim:
+            td_list = []
+            for lazy_td in list_of_tensordicts:
+                td_list += lazy_td.tensordicts
+            out.tensordicts = td_list
```

Comment:
If we have cached values this will break (maybe not now but in the future)
What about `out.append`?

Reply:
we first need to clear out.tensordicts right?
cause when we are here we know that out will have `len(tensordicts) == sum(len(td.tensordicts) for td in list_of_tensordicts) ` 

could you scketch out what you have in mind?

Reply:
you're right but the thing is that this is risky as we change the content of the list without checking what we're doing.
Also thinking about it: this isn't really using `out`, since we discard the data. We should update each tensor in-place, not allocate new tensors

Reply:
okay so something like
```
 out.tensordicts[i].update( torch.cat(
                    [lazy_td.tensordicts[i] for lazy_td in list_of_tensordicts],
                    new_dim,
                ))
```
?